src/cmap/connection_pool.ts

```diff
@@ -354,17 +355,33 @@ export class ConnectionPool extends TypedEventEmitter<ConnectionPoolEvents> {
    * will be held by the pool. This means that if a connection is checked out it MUST be checked back in or
    * explicitly destroyed by the new owner.
    */
-  async checkOut(): Promise<Connection> {
+  async checkOut(options?: { timeout?: Timeout }): Promise<Connection> {
     this.emitAndLog(
       ConnectionPool.CONNECTION_CHECK_OUT_STARTED,
       new ConnectionCheckOutStartedEvent(this)
     );
 
     const waitQueueTimeoutMS = this.options.waitQueueTimeoutMS;
+    const serverSelectionTimeoutMS = this[kServer].topology.s.serverSelectionTimeoutMS;
 
     const { promise, resolve, reject } = promiseWithResolvers<Connection>();
 
-    const timeout = Timeout.expires(waitQueueTimeoutMS);
+    let timeout: Timeout;
+    if (options?.timeout) {
+      // CSOT enabled
+      // Determine if we're using the timeout passed in or a new timeout
+      if (
+        csotMin(options.timeout.duration, serverSelectionTimeoutMS) === serverSelectionTimeoutMS
```

Comment:
Bailey had the opposing opinion: that it should be a standalone helper because its only purpose is to implement CSOT logic.

Re the naming, I'm not sure we'll end up with a satisfactorily descriptive name for this without it being long and unwieldy. How do we feel about just adding a detailed doc comment?